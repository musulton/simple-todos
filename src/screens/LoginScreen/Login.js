import {login} from "../../shared/store/login/LoginAction";
import {useDispatch} from "react-redux";
import {goToScreen} from "../../navigation/NavigationHelper";
import {TODO_PATH} from "../../navigation/NavigationPath";
import {showError, showLoading} from "../../shared/store/app/AppAction";

export const Login = (service) => {
    const dispatch = useDispatch();
    const {callLoginService} = service();
    const onAuthenticate = async (email, password) => {
        try {
            dispatch(showLoading(true));
            await callLoginService(email, password);
            dispatch(login({isLoggedIn: true}));
            goToScreen(TODO_PATH, true);
        } catch (e) {
            dispatch(showError(e.message));
        } finally {
            dispatch(showLoading(false));
        }
    }

    const onDismissError = () => dispatch(showError(''));
    return {
        onAuthenticate,
        onDismissError
    };
};


