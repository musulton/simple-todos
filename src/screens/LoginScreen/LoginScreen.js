import React, {useEffect, useState} from 'react';
import {SafeAreaView, StyleSheet, View,} from 'react-native';
import {useDispatch, useSelector} from "react-redux";

import Input from "../../shared/components/Input";
import SubmitButton from "../../shared/components/SubmitButton";
import {TODO_PATH} from "../../navigation/NavigationPath";
import {showLoading} from "../../shared/store/todo/ToDoAction";
import {login} from "../../shared/store/login/LoginAction";
import Heading from "../../shared/components/Heading";

const LoginScreen = ({navigation}) => {
    const dispatch = useDispatch();
    const [userName, setUserName] = useState('');
    const [password, setPassword] = useState('');
    const isLoggedIn = useSelector((state) => state.LoginReducer.isLoggedIn);

    useEffect(() => {
        if (isLoggedIn) {
            dispatch(showLoading(false));
            navigation.replace(TODO_PATH);
        }
    }, [isLoggedIn])
    const submitLogin = () => {
        dispatch(showLoading(true));

        setTimeout(function () {
            if (userName === '123' && password === '123') {
                dispatch(login(true));
            } else {
                dispatch(login(false));
            }

            dispatch(showLoading(false));
        }, 1000);
    }
    return (
        <SafeAreaView style={styles.container}>
            <View style={styles.content}>
                <View style={{paddingBottom: 20}}>
                    <Heading title={'Todo'}/>
                </View>
                <Input placeholder={"Enter your username"} inputValue={userName} onInputChange={setUserName}/>
                <Input placeholder={"Enter your password"} isSecureText={true} inputValue={password} onInputChange={setPassword}/>
                <SubmitButton title={"Submit"} onSubmit={submitLogin}/>
            </View>
        </SafeAreaView>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    content: {
        flex: 1,
        justifyContent: 'center'
    },
    loading: {
        flex: 1,
        position: 'absolute',
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
        opacity: 0.5,
        backgroundColor: 'black',
        justifyContent: 'center',
        alignItems: 'center'
    }
});

export default LoginScreen;
