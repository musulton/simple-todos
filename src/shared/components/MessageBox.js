import React from 'react'
import {Alert} from 'react-native'

const MessageBox = (title, message, okCallBack) => {
    const showAlert = () => {
        Alert.alert(
            title,
            message,
            [
                {text: "OK", onPress: () => okCallBack()}
            ]
        );
    }
    return {
        showAlert
    }
}

export default MessageBox
