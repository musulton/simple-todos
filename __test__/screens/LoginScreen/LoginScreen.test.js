import {render, screen} from "@testing-library/react-native";

import LoginScreen from "../../../src/screens/LoginScreen/LoginScreen";
import {useSelector} from "react-redux";
import MessageBox from "../../../src/shared/components/MessageBox";

const mockShowAlert = jest.fn();

jest
    .mock("react-redux", () => ({
        ...jest.requireActual("react-redux"),
        useSelector: jest.fn()
    }))
    .mock("../../../src/shared/components/MessageBox", () => ({
        __esModule: true,
        default: jest.fn(() => ({
            showAlert: mockShowAlert
        }))
    }))

describe("LoginScreen", () => {
    const mockAuthenticate = jest.fn();
    const mockDismissError = jest.fn();
    const mockLogin = () => ({
        onAuthenticate: mockAuthenticate,
        onDismissError: mockDismissError
    })

    test("should return correctly", () => {
        render(<LoginScreen login={mockLogin} />)

        expect(screen.toJSON()).toMatchSnapshot()
    })

    test("should show message box when error app is true", () => {
        useSelector.mockImplementation(() => true)

        render(<LoginScreen login={mockLogin} />)

        expect(MessageBox).toHaveBeenCalledWith("Error", true, expect.any(Function))
        expect(mockShowAlert).toHaveBeenCalled()
    })
})
