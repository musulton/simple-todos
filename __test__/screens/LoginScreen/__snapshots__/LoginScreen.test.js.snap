// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen should return correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "paddingBottom": 20,
        }
      }
    >
      <View
        style={
          {
            "marginTop": 40,
          }
        }
      >
        <Text
          style={
            {
              "color": "rgba(175, 47, 47, 0.25)",
              "fontSize": 72,
              "fontWeight": "100",
              "textAlign": "center",
            }
          }
        >
          todos
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#ffffff",
          "elevation": 3,
          "marginLeft": 20,
          "marginRight": 20,
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 3,
            "width": 3,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 3,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter your email"
        placeholderTextColor="#CACACA"
        selectionColor="#666666"
        style={
          {
            "backgroundColor": "#ffffff",
            "height": 60,
            "paddingLeft": 10,
            "paddingRight": 10,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        {
          "backgroundColor": "#ffffff",
          "elevation": 3,
          "marginLeft": 20,
          "marginRight": 20,
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 3,
            "width": 3,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 3,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter your password"
        placeholderTextColor="#CACACA"
        selectionColor="#666666"
        style={
          {
            "backgroundColor": "#ffffff",
            "height": 60,
            "paddingLeft": 10,
            "paddingRight": 10,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        {
          "alignItems": "flex-end",
        }
      }
    >
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "rgba(0,0,0,.1)",
            "borderWidth": 1,
            "height": 50,
            "justifyContent": "center",
            "marginRight": 20,
            "marginTop": 15,
            "paddingLeft": 20,
            "paddingRight": 20,
            "width": 200,
          }
        }
      >
        <Text
          style={
            {
              "color": "#666666",
              "fontWeight": "600",
            }
          }
        >
          Submit
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
